feat: Add Advanced Network Media operators and applications for SMPTE 2110 compliant streaming

ronyrad opened the pull request:
This PR adds multimedia network operators that ease development of media streaming applications with Holoscan Holohub. Two sample applications (sender and player) are provided to demonstrate how to utilize these operators effectively.

### New Components Added

#### 🔧 Core Operators
- **AdvNetworkMediaRxOp**: Media reception operator for receiving streams over Rivermax
  - Packet-to-frame conversion with ANO Burst processing
  - Output as VideoBuffer or Tensor entities
- **AdvNetworkMediaTxOp**: Media transmission operator for sending streams over Rivermax
  - VideoBuffer/Tensor input processing for network transmission
  - Configurable network interface and queue management

#### 📺 Sample Applications
- **Advanced Networking Media Sender**: Demonstrates media streaming
  - Pipeline: `VideoStreamReplayer → AdvNetworkMediaTxOp → Network Interface`
  - Shows frame-accurate timing, looping playback, and file-based sources
- **Advanced Networking Media Player**: Demonstrates media reception
  - Pipeline: `Network Interface → AdvNetworkMediaRxOp → HolovizOp/FramesWriter`
  - Shows real-time visualization, file output, and format conversion

### Developer Benefits
- ✅ **Simplified API** for SMPTE 2110 compliant applications
- ✅ **Easy integration** with existing Holoscan pipelines
- ✅ **Multiple video formats** (RGB888, YUV420, NV12, RGBA) with configurable bit depths (8-16 bit)
- ✅ **Dual language support** (C++ for performance, Python for ease of development)
- ✅ **GPU acceleration** with GPUDirect for zero-copy operations
- ✅ **4K+ resolution support** for high-quality media streaming

### Dependencies
This PR builds upon the foundational Rivermax infrastructure provided by [PR #813](https://github.com/nvidia-holoscan/holohub/pull/813), which includes Rivermax Dev Kit migration and Tx support implementation.

### Target Use Cases
Broadcast and media production environments requiring precise timing and high throughput over IP networks. The sample applications serve as reference implementations for developers building similar streaming solutions.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added two Advanced Networking media apps (player & sender) with C++ and Python runtimes, runnable examples, and options for visualization or writing frames to file.
  * New RX and TX operators enabling end-to-end packet→frame assembly and media transmit/replay pipelines with Python bindings.

* **Enhancements**
  * Adaptive burst-pool capacity monitoring and configurable adaptive-dropping to improve resilience under load.
  * Runtime memory-copy strategy selection (contiguous vs. strided) for optimized zero-copy/host-device paths.
  * Expanded categorized logging and detailed runtime diagnostics.

* **Documentation**
  * Comprehensive READMEs, example YAML configs, and metadata for apps and operators.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ronyrad committed:
feat(advanced_network_media) Add Advanced Network Media operators

- Add AdvNetworkMediaRxOp for receiving media streams over Rivermax
  - Supports packet-to-frame conversion with ANO Burst processing
  - Configurable video formats, bit depths, and frame dimensions
  - Output as VideoBuffer or Tensor entities

- Add AdvNetworkMediaTxOp for transmitting media streams over Rivermax
  - Processes VideoBuffer/Tensor inputs for network transmission
  - Configurable network interface and queue management
  - Support for various video formats

Features:
  - SMPTE 2110 compliance for professional media over IP
  - GPU acceleration and GPUDirect support
  - Low latency optimizations
  - Python bindings for both operators

These operators build upon the Advanced Network library to provide
specialized functionality for broadcast and media streaming
use cases requiring strict timing and high throughput performance.

Signed-off-by: Rony Rado <rrado@nvidia.com>

ronyrad committed:
feat(adv_networking_media_sender) Add Adv. Networking Media Sender

- Add high-performance media streaming application built on Advanced Network Media Tx operator
  - Utilizes AdvNetworkMediaTxOp for media transmission over Rivermax
  - Integrates with Advanced Network Manager for optimized network resource management
  - SMPTE 2110 compliant for professional broadcast applications
  - Real-time transmission of media files with precise timing control

- Support multiple video formats and resolutions
  - RGB888, YUV420, NV12 and other common video formats
  - Configurable bit depths (8, 10, 12, 16-bit)
  - Multiple resolution support up to 4K and beyond
  - GPU acceleration with GPUDirect for zero-copy operations

* Implement dual language support
  - C++ implementation for maximum performance
  - Python implementation for ease of use and development

* Support professional broadcast features
  - Frame-accurate timing for live streaming applications
  - Looping playback for continuous streaming
  - VideoStreamReplayer integration for file-based sources

Pipeline: VideoStreamReplayer → AdvNetworkMediaTxOp → Advanced Network Manager → Network Interface

This application demonstrates how to use the Advanced Network Media operators
for professional-grade media transmission in broadcast and media production
environments requiring ultra-low latency and high throughput performance.

Signed-off-by: Rony Rado <rrado@nvidia.com>

ronyrad committed:
feat(adv_networking_media_player) Add Adv. Networking Media Player

- Add high-performance media receiving application built on Advanced Network Media Rx operator
  - Utilizes AdvNetworkMediaRxOp for professional media reception over Rivermax
  - Integrates with Advanced Network Manager for optimized network resource management
  - SMPTE 2110 compliant for professional broadcast applications
  - Real-time reception and processing of media streams with ultra-low latency

- Support flexible output modes and format handling
  - Real-time visualization using HolovizOp for live monitoring
  - File output capability for recording and analysis
  - Format conversion support for optimal display

- Support multiple video formats and resolutions
  - RGB888, YUV420, NV12, RGBA and other common video formats
  - Configurable bit depths (8, 10, 12, 16-bit)
  - Multiple resolution support up to 4K and beyond
  - GPU acceleration with GPUDirect for zero-copy operations

- Implement dual language support
  - C++ implementation for maximum performance
  - Python implementation for ease of use and development

Pipeline: Network Interface → Advanced Network Manager → AdvNetworkMediaRxOp → [FormatConverter] → HolovizOp/FramesWriter

This application demonstrates how to use the Advanced Network Media operators
for professional-grade media reception in broadcast and media production
environments requiring ultra-low latency and high throughput performance.

Signed-off-by: Rony Rado <rrado@nvidia.com>

agirault commented:
@grlee77 if/when you have any BW, this link would help reviewing just the proper commits, as this is based on #813  : https://github.com/ronyrad/holohub/compare/ano_rivermax_update_mar_25/adv_network_rivermax_manager_update_pr_candidate_01...ronyrad:holohub:ano_rivermax_update_mar_25/adv_network_media_operators_dev_01

ronyrad committed:
feat(advanced_network) Implement adaptive burst pool management with configurable thresholds

- Add adaptive burst dropping system for Rivermax burst pool management
  - Implements capacity-aware burst dropping to prevent memory pool exhaustion
  - Configurable thresholds for warning (25%), critical (10%), and recovery (50%) levels
  - Critical threshold mode drops new bursts when pool capacity falls below threshold
  - Automatic recovery when pool capacity returns to healthy levels
  - Add comprehensive pool capacity monitoring and statistics
  - Extend YAML configuration support for burst pool adaptive dropping
    - New "burst_pool_adaptive_dropping" configuration section
    - Configurable enabled/disabled flag and threshold percentages

This enhancement prevents memory pool exhaustion in high-throughput ANO Rivermax
operations by intelligently dropping incoming bursts when system capacity
is critically low, ensuring system stability while maintaining optimal
performance under normal operating conditions.

Signed-off-by: Rony Rado <rrado@nvidia.com>

ronyrad committed:
feat(advanced_network) Lint fixes

Lint fixes

Signed-off-by: Rony Rado <rrado@nvidia.com>

ronyrad committed:
feat(advanced_network) Update Dockerfile

- Install holoscan python SDK
- Install libvulkan for holoviz operator
- Install xvfb and x11vnc for headless servers

Signed-off-by: Rony Rado <rrado@nvidia.com>

ronyrad commented on operators/advanced_network/advanced_network/managers/rivermax/rivermax_mgr_impl/burst_manager.cpp:
```diff
@@ -394,28 +403,191 @@ RxBurstsManager::RxBurstsManager(bool send_packet_ext_info, int port_id, int que
+      last_capacity_warning_time_ = now;
+    }
+
+    return should_drop_burst_adaptive(utilization);
```
fixed

ronyrad commented on operators/advanced_network/advanced_network/managers/rivermax/rivermax_mgr_impl/burst_manager.cpp:
```diff
@@ -394,28 +403,191 @@ RxBurstsManager::RxBurstsManager(bool send_packet_ext_info, int port_id, int que
+      last_capacity_warning_time_ = now;
+    }
+
+    return should_drop_burst_adaptive(utilization);
```
fixed

ronyrad commented on operators/advanced_network/advanced_network/managers/rivermax/rivermax_mgr_impl/burst_manager.cpp:
```diff
@@ -394,28 +403,191 @@ RxBurstsManager::RxBurstsManager(bool send_packet_ext_info, int port_id, int que
+    }
+
+    // In critical mode, use adaptive dropping but be more aggressive
+    return should_drop_burst_adaptive(utilization);
```
fixed

ronyrad commented on applications/adv_networking_media_player/python/adv_networking_media_player.py:
```diff
@@ -0,0 +1,227 @@
+            )
+
+            self.add_flow(adv_net_media_rx, visualizer, {("out_video_buffer", "receivers")})
+
```
The Holoscan SDK Python API expects add_flow to receive a set of tuples, not a dictionary.
Correct syntax (reverted to):
self.add_flow(adv_net_media_rx, visualizer, {("out_video_buffer", "receivers")})
Why it's correct:
Python uses sets of tuples: {("output", "input")} creates a set containing one tuple
Matches C++ API: C++ uses std::set<std::pair<std::string, std::string>>
Supports multiple connections: {("out1", "in1"), ("out2", "in2")}

bhashemian commented:
@ronyrad could you please take a look at the conflicting changes and resolve it? Thanks

danal97 committed:
[OPERATORS] Convert static variables to instance members in AdvNetworkMediaTxOpImpl

Signed-off-by: Dana Lessner <dlessner@nvidia.com>

danal97 commented on operators/advanced_network_media/advanced_network_media_rx/media_frame_assembler.cpp:
gap is positive, moreover, the if condition checks that gap > kRtpSequenceGapWraparoundThreshold which is a positive number

danal97 committed:
Apply code quality improvements from static analysis review

Signed-off-by: Dana Lessner <dlessner@nvidia.com>

danal97 committed:
Update copyright year to 2025-2026 in various files

Signed-off-by: Dana Lessner <dlessner@nvidia.com>

danal97 committed:
Fix Rivermax build errors

The shared manager.cpp and manager.h depend on DPDK headers so the
Rivermax build target requires DPDK to be present.

- Dockerfile - change the rivermax stage to extend from dpdk instead of
  base-deps.
- manager.cpp DPDK include guard only covered DPDK/GPUNetIO/RDMA builds.
  Extend it to cover ANO_MGR_RIVERMAX as well.
- RivermaxMgr::get_segment_packet_length() and get_packet_length() had
  uint16_t return types, mismatching the uint32_t declared in the
  Manager base class.

Signed-off-by: Dana Lessner <dlessner@nvidia.com>

bhashemian reviewed (commented):
@ronyrad I'd appreciate it if you could take a look at these comments? Thanks

bhashemian commented:
This PR is outdated and no longer align with the changes in Holohub. Advanced Networking has be replaced by DAQIRI(https://github.com/NVIDIA/daqiri).
